Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,205 @@
+
+## Main Definitions
+
+- `Auction`: Defines an auction with bidders and their valuations.
+- `maxb`: Function that computes the highest bid given a bidding function.
+- `winner`: Identifies the winner of the auction as the bidder with the highest bid.
+- `B`: Function that computes the highest bid excluding a given participant.
+- `dominant`: Defines a dominant strategy in the context of a auction.
+- `Secondprice`: Computes the second highest bid in the auction.
+- `Secondprice.utility`: Computes the utility of each bidder based on the outcome of the auction.
+- `Firstprice.utility`: Computes the utility for a first price auction.
+
+## Main Results
+
+- `utility_nneg`: Utility is non-negative if the bid equals the valuation in second price auction.
+- `valuation_is_dominant`: Bidding one's valuation is a dominant strategy in second price auction.
+- `first_price_has_no_dominant_strategy`: There is no dominant strategy in first price auction.
+
+## Helping Lemmas
+
+- `gt_wins`: If `i`'s bid is higher than all other bids, then `i` wins.
+- `exists_max`: There exists a participant whose bid matches the highest bid
+- `winner_take_max`: The winner's bid is the highest.
+- `b_winner`: The winner's bid is at least the second highest bid.
+- `Secondprice.utility_winner`: If `i` wins, utility is the valuation minus the second highest bid.
+- `Secondprice.utility_loser`: If `i` does not win, their utility is 0.
+- `b_winner_max`: The winner's bid is greater than or equal to all other bids.
+- `b_loser_max`: If `i` does not win, the highest bid excluding `i` matches the highest bid.
+- `Firstprice.utility_winner`: If `i` wins in a first price auction utility is valuation minus bid.
+- `Firstprice.utility_loser`: If `i` does not win in a first price auction, utility is 0.
+
+## Implementation Notes
+
+The structure and functions assume the existence of multiple bidders to allow for meaningful
+auction dynamics. Definitions like `winner` and `maxb` make use of Lean's `Finset` and `Classical`
+logic to handle potential non-constructive cases effectively.
+
+## References
+
+* [T. Roughgarden, *Twenty lectures on Algorithmic Game Theory*][roughgarden2016]
+
+## Tags
+
+auction, game theory, economics, bidding, valuation
+-/
+
+open Classical
+
+/-!### Structure Definition -/
+
+/-- The Auction structure is set with components including: -/
+structure Auction where
+   /-- A set of participants.-/
+   I : Type*
+   /-- A `Fintype` instance. -/
+   hF : Fintype I
+   /-- An `Inhabited` instance. -/
+   hI : Nontrivial I
+   /-- A function mapping each participant to their valuation. -/
+   v : I → ℝ
+
+attribute [instance] Auction.hF Auction.hI
+
+namespace Auction
+
+variable {a : Auction} (b : a.I → ℝ)
+
+/-!### Helper Functions and Definitions -/
+/-- Computes the highest bid given a bidding function `b`. -/
+@[simp]
+def maxb : ℝ := Finset.sup' Finset.univ Finset.univ_nonempty b
```

Comment:
Some of these names will need to be changed. For instance the names `maxb`, `b_winner_max`, `b_winner`, `b_winner_le_maxb`, and `b_loser_max` make reference to the variable name `b`, which is not ideal since the meaning will not be clear to someone seeing these names from outside the file. Here's one list of possibilities, but feel free to think about other options.
`maxb` -> `maxBid`
`exists_max` -> `exists_maxBad` or `exists_eq_maxBid` or `exists_bid_eq_maxBid`
`winner_take_max` -> `bid_winner` or `bid_winner_eq_maxBid`
`b_winner_max` -> `bid_le_bid_winner`
`gt_wins` -> `eq_winner_of_bid_gt` (or similar; I'm not sure what's best here)